### PR TITLE
Add ServiceClaim status condition

### PR DESCRIPTION
**Pull request by baijum**
Jira: https://issues.redhat.com/browse/APPSVC-1296

### CODE REVIEW

File: controllers/agents/app/serviceclaim_controller.go [hunk 1]
```diff
@@ -134,6 +137,15 @@ func (r *ServiceClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	})
 	if err != nil {
 		if strings.Contains(err.Error(), "admission webhook \"vserviceclaim.kb.io\" denied the request") {
+			c := metav1.Condition{
+				LastTransitionTime: metav1.NewTime(time.Now()),
+				Type:               primazaiov1alpha1.ServiceBindingConditionReady,
+				Status:             metav1.ConditionFalse,
+				Reason:             "WebhookError",
+				Message:            "",
```
Comment (Kartikey-star) [inline controllers/agents/app/serviceclaim_controller.go:145]: Can we use `err.Error()` in Message

File: controllers/agents/app/serviceclaim_controller.go [hunk 2]
```diff
@@ -134,6 +137,15 @@ func (r *ServiceClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	})
 	if err != nil {
 		if strings.Contains(err.Error(), "admission webhook \"vserviceclaim.kb.io\" denied the request") {
+			c := metav1.Condition{
+				LastTransitionTime: metav1.NewTime(time.Now()),
+				Type:               primazaiov1alpha1.ServiceBindingConditionReady,
```
Comment (Kartikey-star) [inline controllers/agents/app/serviceclaim_controller.go:142]: Should be `ServiceClaimConditionReady` and not `ServiceBindingConditionReady`. Applies in all places. Do you think we can created three types for different states?

File: controllers/serviceclaim_controller.go [hunk 3]
```diff
@@ -301,6 +302,15 @@ func (r *ServiceClaimReconciler) processServiceClaim(
 	}
 
 	if !registeredServiceFound {
+		c := metav1.Condition{
```
Comment (Kartikey-star) [inline controllers/serviceclaim_controller.go:305]: Let's create a method which updates the service claim passing in parameters like status,reason,message and state and update the status of service claim too.And use that wherever required.

File: controllers/agents/app/serviceclaim_controller.go (hunk 1 above)
Comment (filariow) [inline controllers/agents/app/serviceclaim_controller.go:145]: I'd also propose to use a less technical Reason here

```suggestion
				Reason:             "ValidationError",
				Message:            err.Error(),
```

File: controllers/serviceclaim_controller.go [hunk 4]
```diff
@@ -301,6 +302,15 @@ func (r *ServiceClaimReconciler) processServiceClaim(
...
+			LastTransitionTime: metav1.NewTime(time.Now()),
+			Type:               primazaiov1alpha1.ServiceBindingConditionReady,
+			Status:             metav1.ConditionFalse,
+			Reason:             "UnmatchedSCI",
+			Message:            "",
```
Comment (filariow) [inline controllers/serviceclaim_controller.go:310]: I'd propose to use a less technical reason 

```suggestion
			Reason:             "NoMatchingServiceFound",
			Message:            err.Error(),
```

File: controllers/serviceclaim_controller.go [hunk 5]
```diff
@@ -313,6 +323,15 @@ func (r *ServiceClaimReconciler) processServiceClaim(
 	// if the number of SED keys is more than the number of secret data entries
 	// that indicates one or more keys are missing
 	if len(sclaim.Spec.ServiceEndpointDefinitionKeys) > count {
+		c := metav1.Condition{
+			LastTransitionTime: metav1.NewTime(time.Now()),
+			Type:               primazaiov1alpha1.ServiceBindingConditionReady,
+			Status:             metav1.ConditionFalse,
+			Reason:             "MissingKeyInSEDs",
+			Message:            "",
```
Comment (filariow) [inline controllers/serviceclaim_controller.go:331]: I'd propose to use a less technical reason 

```suggestion
			Reason:             "NoMatchingServiceFound",
			Message:            err.Error(),
```

**Review by filariow — CHANGES_REQUESTED** [review]
some proposals

File: controllers/agents/app/serviceclaim_controller.go (hunk 2 above)
Comment (baijum) [reply controllers/agents/app/serviceclaim_controller.go:142]: I think a common type make the condition consistent.

File: controllers/agents/app/serviceclaim_controller.go (hunk 1 above)
Comment (baijum) [reply controllers/agents/app/serviceclaim_controller.go]: Done

Comment (baijum) [reply controllers/agents/app/serviceclaim_controller.go]: The error object is not available. But I have updated the message. Please check.

File: controllers/serviceclaim_controller.go (hunk 4 above)
Comment (baijum) [reply controllers/serviceclaim_controller.go]: The error object is not available. But I have updated the message. Please check.

**Review by filariow — APPROVED** [review]
lgtm, just an optional proposal

File: controllers/agents/app/serviceclaim_controller.go [hunk 6]
```diff
@@ -134,6 +137,15 @@ func (r *ServiceClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	})
 	if err != nil {
 		if strings.Contains(err.Error(), "admission webhook \"vserviceclaim.kb.io\" denied the request") {
+			c := metav1.Condition{
+				LastTransitionTime: metav1.NewTime(time.Now()),
+				Type:               primazaiov1alpha1.ServiceClaimConditionReady,
+				Status:             metav1.ConditionFalse,
+				Reason:             "ValidationError",
```
[the remark this answers is not in the document]
Comment (baijum) [reply controllers/agents/app/serviceclaim_controller.go]: Done.

File: controllers/serviceclaim_controller.go (hunk 3 above)
Comment (baijum) [reply controllers/serviceclaim_controller.go:306]: Here we are initializing a struct and calling a method. I don't know if it's a good idea to create function for the same.

**Review by Kartikey-star — APPROVED** [review]

**Review by filariow — APPROVED** [review]
lgtm

**Pull request merged**